File: experiments/experiment_jobs/ingestion/src/main/scala/ingest_data.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.functions._

import java.time.{Duration, Instant, LocalDate, LocalDateTime}
import java.time.format.DateTimeFormatter

object ingest_data {

  case class JobConfig(
                        jdbc_url: String,
                        source_db_user: String,
                        source_db_password: String,
                        source_table: String,
                        target_table: String,
                        ingestion_time: String
                      )

  private val job_name = "ingest_data"

  def parse_args(args: Array[String]): JobConfig = {
    JobConfig(
      jdbc_url = get_required_arg(args, "jdbc_url"),
      source_db_user = get_required_arg(args, "source_db_user"),
      source_db_password = get_required_arg(args, "source_db_password"),
      source_table = get_required_arg(args, "source_table"),
      target_table = get_required_arg(args, "target_table"),
      ingestion_time = get_optional_arg(args, "ingestion_time", "NONE")
    )
  }

  private def get_required_arg(args: Array[String], name: String): String = {
    val index = args.indexOf(s"--${name}")

    if (index == -1 || index + 1 >= args.length) {
      throw new IllegalArgumentException(s"Missing required argument: --${name}")
    }

    args(index + 1)
  }

  private def get_optional_arg(args: Array[String], name: String, default_value: String): String = {
    val index = args.indexOf(s"--${name}")

    if (index == -1 || index + 1 >= args.length) {
      default_value
    } else {
      args(index + 1)
    }
  }

  private def build_spark_session(config: JobConfig): SparkSession = {
    SparkSession.builder()
      .appName(job_name)
      .getOrCreate()
  }

  private def read_table_from_data_source(
                                           spark: SparkSession,
                                           job_config: JobConfig
                                         ): DataFrame = {
    val dbtable = build_source_query(job_config)

    spark.read
      .format("jdbc")
      .option("url", job_config.jdbc_url)
      .option("dbtable", dbtable)
      .option("user", job_config.source_db_user)
      .option("password", job_config.source_db_password)
      .option("driver", "org.postgresql.Driver")
      .option("fetchsize", "20000")
      .load()
  }

  private def build_source_query(job_config: JobConfig): String = {
    if (job_config.ingestion_time == "NONE") {
      return job_config.source_table
    }

    val start_time = get_start_time(job_config.ingestion_time)
    val end_time = get_end_time(job_config.ingestion_time, start_time)

    val start_time_sql = start_time.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
    val end_time_sql = end_time.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))

    s"""
       |(
       |  SELECT *
       |  FROM ${job_config.source_table}
       |  WHERE date_hour >= TIMESTAMP '${start_time_sql}'
       |    AND date_hour <  TIMESTAMP '${end_time_sql}'
       |) AS source_data
       |""".stripMargin
  }

  private def get_start_time(ingestion_time: String): LocalDateTime = {
    if (is_date_only(ingestion_time)) {
      val date = LocalDate.parse(
        ingestion_time,
        DateTimeFormatter.ofPattern("yyyy-MM-dd")
      )

      return date.atStartOfDay()
    }

    LocalDateTime.parse(
      ingestion_time,
      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
    )
  }

  private def get_end_time(
                            ingestion_time: String,
                            start_time: LocalDateTime
                          ): LocalDateTime = {
    if (is_date_only(ingestion_time)) {
      return start_time.plusDays(1)
    }

    start_time.plusHours(1)
  }

  private def is_date_only(value: String): Boolean = {
    value.matches("\\d{4}-\\d{2}-\\d{2}")
  }

  private def convert_column_names_to_lower_case(df: DataFrame): DataFrame = {
    val new_columns = df.columns.map { column_name =>
      col(column_name).as(column_name.toLowerCase)
    }

    df.select(new_columns: _*)
  }

  private def transform_data(df: DataFrame): DataFrame = {
    convert_column_names_to_lower_case(df)
  }

  private def write_to_iceberg(df: DataFrame, target_table: String): Unit = {
    df.write
      .format("iceberg")
      .mode("append")
      .save(s"iceberg.${target_table}")
  }

  private def start_timer(step_name: String): Instant = {
    val start_time = Instant.now()
    println(s"[Performance] START - ${step_name} at ${start_time}")
    start_time
  }

  private def end_timer(step_name: String, start_time: Instant): Unit = {
    val end_time = Instant.now()
    val duration_ms = Duration.between(start_time, end_time).toMillis
    val duration_seconds = duration_ms / 1000.0

    println(s"[Performance] END   - ${step_name} at ${end_time}")
    println(s"[Performance] TIME  - ${step_name}: ${duration_ms} ms (${duration_seconds} seconds)")
  }

  private def log_job_config(job_config: JobConfig): Unit = {
    println("[Experiment] Starting batch ingest job")
    println(s"Source table: ${job_config.source_table}")
    println(s"Target table: ${job_config.target_table}")
    println(s"Ingestion time: ${job_config.ingestion_time}")
  }

  def main(args: Array[String]): Unit = {
    val job_start_time = Instant.now()
    println(s"[Performance] JOB START at ${job_start_time}")

    val parse_start_time = start_timer("Parse arguments")
    val job_config = parse_args(args)
    end_timer("Parse arguments", parse_start_time)

    val spark_start_time = start_timer("Build Spark session")
    val spark = build_spark_session(job_config)
    end_timer("Build Spark session", spark_start_time)

    try {
      log_job_config(job_config)

      val read_plan_start_time = start_timer("Create PostgreSQL DataFrame")
      val source_df = read_table_from_data_source(
        spark = spark,
        job_config = job_config
      )
      end_timer("Create PostgreSQL DataFrame", read_plan_start_time)

      val transform_plan_start_time = start_timer("Create transform DataFrame")
      val transformed_df = transform_data(source_df)
      end_timer("Create transform DataFrame", transform_plan_start_time)

      val write_start_time = start_timer("Write data to Iceberg")
      write_to_iceberg(
        df = transformed_df,
        target_table = job_config.target_table
      )
      end_timer("Write data to Iceberg", write_start_time)

      val job_end_time = Instant.now()
      val total_duration_ms = Duration.between(job_start_time, job_end_time).toMillis
      val total_duration_seconds = total_duration_ms / 1000.0

      println(s"[Performance] JOB END at ${job_end_time}")
      println(s"[Performance] TOTAL TIME: ${total_duration_ms} ms (${total_duration_seconds} seconds)")
      println("Batch ingestion job completed successfully")

    } finally {
      val stop_start_time = start_timer("Stop Spark session")
      spark.stop()
      end_timer("Stop Spark session", stop_start_time)
    }
  }
}

File: datagate/infra/shared/jobs/experiments/pyspark/tlc_trip_record/ingest_data.py
import argparse
from datetime import datetime, timedelta
from time import perf_counter

from pyspark.sql import DataFrame, SparkSession
from pyspark.sql.functions import col, lit, to_timestamp

JOB_NAME = "ingest_data"

def parse_args():
    parser = argparse.ArgumentParser(
        description="Ingest one processing batch from PostgreSQL to Iceberg bronze table"
    )

    parser.add_argument("--jdbc_url", required=True)
    parser.add_argument("--source_db_user", required=True)
    parser.add_argument("--source_db_password", required=True)
    parser.add_argument("--source_table", required=True)
    parser.add_argument("--target_table", required=True)
    parser.add_argument("--processing_date_hour", required=True)
    parser.add_argument("--spark_master", default="spark://spark-master:7077")
    parser.add_argument("--spark_driver_cores", type=int, default=2)
    parser.add_argument("--spark_driver_memory", default="4g")
    parser.add_argument("--spark_executor_instances", type=int, default=2)
    parser.add_argument("--spark_executor_cores", type=int, default=6)
    parser.add_argument("--spark_executor_memory", default="7g")
    parser.add_argument("--spark_sql_shuffle_partitions", type=int, default=32)
    parser.add_argument("--spark_default_parallelism", type=int, default=32)

    return parser.parse_args()


def build_spark_session(args) -> SparkSession:
    spark = (
        SparkSession.builder
        .appName(JOB_NAME)
        .master(args.spark_master)
        .config("spark.driver.cores", str(args.spark_driver_cores))
        .config("spark.driver.memory", args.spark_driver_memory)
        .config("spark.executor.instances", str(args.spark_executor_instances))
        .config("spark.executor.cores", str(args.spark_executor_cores))
        .config("spark.executor.memory", args.spark_executor_memory)
        .config("spark.sql.shuffle.partitions", str(args.spark_sql_shuffle_partitions))
        .config("spark.default.parallelism", str(args.spark_default_parallelism))
        .config("spark.sql.adaptive.enabled", "true")
        .getOrCreate()
    )

    return spark


def resolve_batch_window(processing_date_hour: str) -> tuple[str, str]:
    batch_end = datetime.strptime(processing_date_hour, "%Y-%m-%d %H:%M:%S")

    if batch_end.hour not in {0, 12} or batch_end.minute != 0 or batch_end.second != 0:
        raise ValueError(
            "processing_date_hour must be in format YYYY-MM-DD HH:MM:SS "
            "and the hour must be 00:00:00 or 12:00:00"
        )

    batch_start = batch_end - timedelta(hours=12)

    batch_start_str = batch_start.strftime("%Y-%m-%d %H:%M:%S")
    batch_end_str = batch_end.strftime("%Y-%m-%d %H:%M:%S")

    return batch_start_str, batch_end_str


def build_source_query(source_table: str, processing_date_hour: str) -> str:
    batch_start, batch_end = resolve_batch_window(processing_date_hour)

    query = f"""
    (
        SELECT *
        FROM {source_table}
        WHERE date_hour >= TIMESTAMP '{batch_start}'
          AND date_hour < TIMESTAMP '{batch_end}'
    ) AS source_data
    """

    return query


def read_table_from_data_source(
    spark: SparkSession,
    jdbc_url: str,
    source_db_user: str,
    source_db_password: str,
    source_table: str,
    processing_date_hour: str,
) -> DataFrame:
    dbtable = build_source_query(
        source_table=source_table,
        processing_date_hour=processing_date_hour,
    )

    df = (
        spark.read
        .format("jdbc")
        .option("url", jdbc_url)
        .option("dbtable", dbtable)
        .option("user", source_db_user)
        .option("password", source_db_password)
        .option("driver", "org.postgresql.Driver")
        .option("fetchsize", "20000")
        .load()
    )

    return df


def convert_column_names_to_lower_case(df: DataFrame) -> DataFrame:
    lower_case_df = df.select(
        *[
            col(column_name).alias(column_name.lower())
            for column_name in df.columns
        ]
    )

    return lower_case_df


def transform_data(df: DataFrame, processing_date_hour: str) -> DataFrame:
    transformed_df = (
        convert_column_names_to_lower_case(df)
        .withColumn(
            "processing_date_hour",
            to_timestamp(lit(processing_date_hour)),
        )
    )

    return transformed_df


def write_to_iceberg(df: DataFrame, target_table: str) -> None:
    (
        df.writeTo(f"iceberg.{target_table}")
        .overwritePartitions()
    )


def main() -> None:
    job_start_time = perf_counter()
    args = parse_args()
    spark = build_spark_session(args)

    try:
        batch_start, batch_end = resolve_batch_window(args.processing_date_hour)

        print(
            f"[{JOB_NAME}] Processing source date_hour window "
            f"[{batch_start}, {batch_end})"
        )

        source_df = read_table_from_data_source(
            spark=spark,
            jdbc_url=args.jdbc_url,
            source_db_user=args.source_db_user,
            source_db_password=args.source_db_password,
            source_table=args.source_table,
            processing_date_hour=args.processing_date_hour,
        )

        transformed_df = transform_data(
            df=source_df,
            processing_date_hour=args.processing_date_hour,
        )

        write_to_iceberg(
            df=transformed_df,
            target_table=args.target_table,
        )

        total_seconds = perf_counter() - job_start_time

        print(
            f"[Job Completed] {JOB_NAME} finished in {total_seconds:.3f} seconds"
        )

    finally:
        spark.stop()


if __name__ == "__main__":
    main()